File: App/MainApp/IntegrationTests/ObservableTestResults.vb
Option Strict On
Option Infer Off
Option Explicit On

Namespace IntegrationTests

    Public Class TestResult
        Public TestName As String
        Public Passed As Boolean
        Public ExceptionMessage As String
        Public StackTrace As List(Of String)
    End Class

    Public Class ObservableTestSuiteResults
        Public TimeElapsed As String
        Public TotalTests As Integer
        Public PassedTests As Integer
        Public FailedTests As Integer
        Public Messages As New List(Of String)
        Public TestResults As New List(Of TestResult)

        Public Event MessageAdded(sender As Object, e As EventArgs)
        Public Sub AddMessage(newItem As String)
            Messages.Add(newItem)
            RaiseEvent MessageAdded(Me, New EventArgs())
        End Sub


        Public Event ResultAdded(sender As Object, e As EventArgs)
        Public Sub AddResult(newItem As TestResult)
            TestResults.Add(newItem)
            RaiseEvent ResultAdded(Me, New EventArgs())
        End Sub

    End Class


End Namespace

File: CSharpApp/MainApp/IntegrationTests/ObservableTestResults.cs
using System;
using System.Collections.Generic;

namespace MainApp.IntegrationTests
{

    public class TestResult
    {
        public string TestName;
        public bool Passed;
        public string ExceptionMessage;
        public List<string> StackTrace;
    }

    public class ObservableTestSuiteResults
    {
        public string TimeElapsed;
        public int TotalTests;
        public int PassedTests;
        public int FailedTests;
        public List<string> Messages = new List<string>();
        public List<TestResult> TestResults = new List<TestResult>();

        public event MessageAddedEventHandler MessageAdded;

        public delegate void MessageAddedEventHandler(object sender, EventArgs e);
        public void AddMessage(string newItem)
        {
            Messages.Add(newItem);
            MessageAdded?.Invoke(this, new EventArgs());
        }


        public event ResultAddedEventHandler ResultAdded;

        public delegate void ResultAddedEventHandler(object sender, EventArgs e);
        public void AddResult(TestResult newItem)
        {
            TestResults.Add(newItem);
            ResultAdded?.Invoke(this, new EventArgs());
        }

    }


}